### PR TITLE
Mnot/2620

### LINKED ISSUE #2620: ABNF issues
In the current ABNF, Display Strings are built upon:

~~~ abnf
uchr             = chr / uescaped
~~~

However, `chr` is:

~~~ abnf
chr       = unescaped / escaped
~~~

I think it would be better to remove the `chr` and `uchr` abstractions and just put the alternations directly in `sf-string` and `sf-displaystring`. 

It'd probably also be good to change `escaped` to `backslash-escaped` and `uescaped` (which is super confusing) to `percent-escaped`.

Also, `\` (`%x5C`) _is_ allowed in display strings.

And HEXDIG is not case sensitive.

Yes, I've been using a fuzzer.

### PR DESCRIPTION
**Pull request by mnot**
Fixes #2620.

### CODE REVIEW

File: draft-ietf-httpbis-sfbis.md [hunk 1]
```diff
@@ -1102,37 +1102,31 @@ key           = ( lcalpha / "*" )
...
-sf-boolean = "?" boolean
-boolean    = "0" / "1"
-
-sf-date = "@" sf-integer
-
-sf-displaystring = "%" DQUOTE *uchr DQUOTE
-uchr             = chr / uescaped
-uescaped         = "%" HEXDIG HEXDIG
+pct-encoded  = "%" lc-hexdig lc-hexdig
+lc-hexdig = DIGIT / "a" / "b" / "c" / "d" / "e" / "f"
```
Comment (martinthomson) [inline draft-ietf-httpbis-sfbis.md:1129]: IMPORTANT: "a" means "a" or "A".  You really need to use either `%x61-66` or the RFC 7405 extension of `%s"a" / %s"b" ` and so on.

Comment (mnot) [reply draft-ietf-httpbis-sfbis.md:1129]: Thanks for the reminder.

File: draft-ietf-httpbis-sfbis.md [hunk 2]
```diff
@@ -1102,37 +1102,31 @@ key           = ( lcalpha / "*" )
...
+sf-decimal       = ["-"] 1*12DIGIT "." 1*3DIGIT
+sf-string        = DQUOTE *( unescaped / "%" / bs-escaped ) DQUOTE
+sf-token         = ( ALPHA / "*" ) *( tchar / ":" / "/" )
+sf-binary        = ":" *base64 ":"
+sf-boolean       = "?" ( "0" / "1" )
+sf-date          = "@" sf-integer
+sf-displaystring = "%" DQUOTE *( unescaped / "\" / pct-encoded ) DQUOTE
 
-sf-decimal  = ["-"] 1*12DIGIT "." 1*3DIGIT
+base64       = ALPHA / DIGIT / "+" / "/" / "="
```
Comment (martinthomson) [inline draft-ietf-httpbis-sfbis.md:1123]: base64 is really:

```abnf
base64 = *(ALPHA / DIGIT / "+" / "/") *"="
```

Comment (mnot) [reply draft-ietf-httpbis-sfbis.md:1123]: Yeah, I was wondering how far to go down that rabbit hole. I think that's a nice balance.

Comment (martinthomson) [reply draft-ietf-httpbis-sfbis.md:1123]: Yup.  `*3"="` is one step too many perhaps.

Comment (mnot) [reply draft-ietf-httpbis-sfbis.md:1123]: In particular, I don't want to imply that we're validating base64 here -- same as the percent encoding.

**Pull request merged**